contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,794 @@
+
+        if (_withdrawalQueue().isPaused()) revert WQNotResumed();
+        if (_validatorsExitBusOracle().isPaused()) revert VEBONotResumed();
+
+        if (!_stakingRouter().hasStakingModule(1) || _stakingRouter().hasStakingModule(2)) {
+            revert WrongStakingModulesCount();
+        }
+    }
+
+    function _assertGateSealSealables() internal view {
+        // TODO: sync VEBO proxy and its sealable at re-deploy
+        address[] memory sealables = IGateSeal(_gateSeal).get_sealables();
+        if (
+            sealables[0] != address(_withdrawalQueue())
+         || sealables[1] != address(_validatorsExitBusOracle())
+         ) {
+            revert WrongSealGateSealables();
+        }
+    }
+
+    function _assertCorrectOracleAndConsensusContractsBinding(IBaseOracle oracle, IHashConsensus hashConsensus) internal view {
+        if (
+            oracle.getConsensusContract() != address(hashConsensus)
+         || hashConsensus.getReportProcessor() != address(oracle)
+        ) {
+            revert IncorrectOracleAndHashConsensusBinding(address(oracle), address(hashConsensus));
+        }
+    }
+
+    function _checkContractVersions() internal view {
+        _assertContractVersion(_lido(), EXPECTED_FINAL_LIDO_VERSION);
+        _assertContractVersion(_nodeOperatorsRegistry, EXPECTED_FINAL_NODE_OPERATORS_REGISTRY_VERSION);
+        _assertContractVersion(_legacyOracle(), EXPECTED_FINAL_LEGACY_ORACLE_VERSION);
+        _assertContractVersion(_accountingOracle(), EXPECTED_FINAL_ACCOUNTING_ORACLE_VERSION);
+        _assertContractVersion(_stakingRouter(), EXPECTED_FINAL_STAKING_ROUTER_VERSION);
+        _assertContractVersion(_validatorsExitBusOracle(), EXPECTED_FINAL_VALIDATORS_EXIT_BUS_ORACLE_VERSION);
+        _assertContractVersion(_withdrawalQueue(), EXPECTED_FINAL_WITHDRAWAL_QUEUE_VERSION);
+        _assertContractVersion(_withdrawalVault(), EXPECTED_FINAL_WITHDRAWAL_VAULT_VERSION);
+    }
+
+    function _assertContractVersion(IVersioned versioned, uint256 expectedVersion) internal view {
+        if (versioned.getContractVersion() != expectedVersion) {
+            revert InvalidContractVersion(address(versioned), expectedVersion);
+        }
+    }
+
+    function _transferOZAdminFromThisToAgent(IAccessControlEnumerable accessControlled) internal {
+        accessControlled.grantRole(DEFAULT_ADMIN_ROLE, _agent);
+        accessControlled.renounceRole(DEFAULT_ADMIN_ROLE, address(this));
+    }
+
+    function _resumeWithdrawalQueue() internal {
+        IWithdrawalQueue wq = _withdrawalQueue();
+        bytes32 resume_role = wq.RESUME_ROLE();
+        wq.grantRole(resume_role, address(this));
+        wq.resume();
+        wq.renounceRole(resume_role, address(this));
+    }
+
+    // To be strict need two almost identical _resume... function because RESUME_ROLE and resume()
+    // do not actually belong to PausableUntil contract
```

Comment:
Not a blocker, though could have introduced "proper" interface to do just that. Third function would have been a no-go.

Reply:
At first there was a IPausableUntil with `resume` and `PAUSE_ROLE` and `RESUME_ROLE`, but @TheDZhon didn't like it, and I agreed, because its more true to display that the roles and `resume` belong actually to multiple contracts, although happen to have same values/signatures. It's more accurate and strict in general. But I agree both options are good.

Reply:
Yeah, now I see that it has both sides exactly. 

Maybe introduce a little helper interface still, not enriching `IPauseble`, but have a dedicated one.